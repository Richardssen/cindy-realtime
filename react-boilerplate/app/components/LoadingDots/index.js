/**
 *
 * LoadingDots
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Flex } from 'rebass';

export const Dot = styled.span`
  padding: ${(props) => props.size || 5}px;
  margin: ${(props) => props.size || 5}px;
  border-radius: 9999px;
  background-color: ${(props) => (props.active ? '#666' : '#aaa')};
`;

class LoadingDots extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      active: 1,
    };
  }

  componentDidMount() {
    this.timer = window.setInterval(
      () =>
        this.setState((prevState) => ({ active: (prevState.active + 1) % 3 })),
      1000
    );
  }

  componentWillUnmount() {
    window.clearInterval(this.timer);
  }

  render() {
    const { size, ...others } = this.props;
    return (
      <Flex justify="center" {...others}>
        <Dot active={this.state.active === 0} size={size} />
        <Dot active={this.state.active === 1} size={size} />
        <Dot active={this.state.active === 2} size={size} />
      </Flex>
    );
  }
}

LoadingDots.propTypes = {
  size: PropTypes.number,
};

export default LoadingDots;
