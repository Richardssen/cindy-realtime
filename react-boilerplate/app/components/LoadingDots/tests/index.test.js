import React from 'react';
import { shallow, mount } from 'enzyme';

import LoadingDots, { Dot } from '../index';

describe('<LoadingDots />', () => {
  describe('<LoadingDots />', () => {
    it('Changing size of LoadingDots', () => {
      const size = 3;
      const rendered = shallow(<LoadingDots size={size} />);
      expect(rendered.find('[active=true]').length).toEqual(1);
      expect(rendered.find('[active=true]').prop('size')).toEqual(size);
    });
    it('Showing animation', () => {
      jest.useFakeTimers();

      // Start Timer
      const rendered = mount(<LoadingDots />);
      expect(rendered.find('[active=true]').length).toEqual(1);
      expect(window.setInterval).toHaveBeenCalledTimes(1);
      const active1 = rendered.state('active');

      // After one timer count
      jest.runOnlyPendingTimers();
      const active2 = rendered.state('active');
      expect(active1).not.toBe(active2);

      // Clear Timer
      rendered.unmount();
      expect(window.clearInterval).toHaveBeenCalledTimes(1);
    });
  });
  describe('<Dot />', () => {
    it('Without Props', () => {
      const rendered = shallow(<Dot />);
      expect(rendered).toMatchSnapshot();
    });
    it('With Props', () => {
      const rendered = shallow(<Dot size={3} active />);
      expect(rendered).toMatchSnapshot();
    });
  });
});
