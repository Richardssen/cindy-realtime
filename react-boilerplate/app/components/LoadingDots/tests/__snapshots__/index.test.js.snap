// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoadingDots /> <Dot /> With Props 1`] = `
<span
  className="LoadingDots__Dot-jTopvB fHIBhv"
  size={3}
/>
`;

exports[`<LoadingDots /> <Dot /> Without Props 1`] = `
<span
  className="LoadingDots__Dot-jTopvB jrttWY"
/>
`;
